src/libraries/Microsoft.Extensions.FileProviders.Physical/tests/PhysicalFileProviderTests.cs

```diff
@@ -87,6 +89,48 @@ public void GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes_U
             GetFileInfoReturnsNotFoundFileInfoForIllegalPathWithLeadingSlashes(path);
         }
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Linux)] 
+        public async void PollingFileProviderShouldntConsumeINotifyInstances()
+        {
+            List<IDisposable> disposables = new List<IDisposable>();
+            using var root = new DisposableFileSystem();
+
+            int callbacks = 0;
+            int maxInstances = int.Parse(File.ReadAllText("/proc/sys/fs/inotify/max_user_instances"));
+            int instances = maxInstances + 16;
+            try
+            {
+                for (int i = 0; i < instances; i++)
+                {
+                    PhysicalFileProvider pfp = new PhysicalFileProvider(root.RootPath)
+                    {
+                        UsePollingFileWatcher = true,
+                        UseActivePolling = true
+                    };
+                    disposables.Add(pfp);
+                    disposables.Add(pfp.Watch("*").RegisterChangeCallback(
+                        o => Interlocked.Increment(ref callbacks),
+                        i));
+                }
+
+                // trigger an event
+                root.CreateFile("test.txt");
+
+                // let some events fire
+                await Task.Delay(WaitTimeForTokenToFire * 16);
+
+                Assert.NotEqual(0, callbacks);
```

Comment:
Instead of using a `int callbacks` count, would it make more sense to use an `AutoResetEvent` (or similar) that gets set in the callback?

Reply:
yep, good suggestion